samples/async_op_samples_int/hl_async_op_module.c

```diff
@@ -0,0 +1,316 @@
+        result = MU_FAILURE;
+    }
+    else
+    {
+        if (sm_exec_begin(handle->sm) != SM_EXEC_GRANTED)
+        {
+            LogError("sm_exec_begin failed");
+            result = MU_FAILURE;
+        }
+        else
+        {
+            // 1. Create an ASYNC_OP for the operation context
+            THANDLE(ASYNC_OP) async_op = async_op_create(on_async_op_HL_ASYNC_OP_MODULE_EXECUTE_CONTEXT_cancel, sizeof(HL_ASYNC_OP_MODULE_EXECUTE_CONTEXT), alignof(HL_ASYNC_OP_MODULE_EXECUTE_CONTEXT), on_async_op_HL_ASYNC_OP_MODULE_EXECUTE_CONTEXT_dispose);
+            if (async_op == NULL)
+            {
+                LogError("async_op_create failed");
+                result = MU_FAILURE;
+            }
+            else
+            {
+                // 2. Any context needed is stored in async_op->context
+                HL_ASYNC_OP_MODULE_EXECUTE_CONTEXT* async_op_context = async_op->context;
+
+                async_op_context->callback = callback;
+                async_op_context->context = context;
+                async_op_context->handle = handle;
+
+                // 3. Initialize the ll_async_op to NULL
+                THANDLE_INITIALIZE(ASYNC_OP)(&async_op_context->ll_async_op, NULL);
+
+                // 4. Store the async_op from the LL in a temporary variable
+                THANDLE(ASYNC_OP) ll_async_op = NULL;
+
+                // 5. Take an additional reference on the async_op, we have 1 reference for returning to the caller (async_op), and 1 reference for the async work callback (async_op_ref_for_callback)
+                //    Need to take the reference before starting the operation below because its callback may come immediately
+                THANDLE(ASYNC_OP) async_op_ref_for_callback = NULL;
+                THANDLE_ASSIGN(ASYNC_OP)(&async_op_ref_for_callback, async_op);
+
+                // 6. Start the async work, passing the async_op_ref_for_callback as the context
+                if (handle->ll_execute_async(handle->ll_handle, complete_in_ms, &ll_async_op, hl_async_op_module_on_ml_complete, (void*)async_op_ref_for_callback) != 0)
+                {
+                    LogError("ll_execute_async for lower module failed");
+                    result = MU_FAILURE;
+                }
+                else
+                {
+                    // 7. Store the ll_async_op in the context on success so that it can be canceled
+                    //    This must happen before the next line so that the ll_async_op is set before cancel is called
+                    THANDLE_MOVE(ASYNC_OP)(&async_op_context->ll_async_op, &ll_async_op);
+
+                    // 8. Provide the async_op to the caller
+                    //    After this (but before returning) it is possible that the caller may call cancel because they may have access to the async_op_out pointer in another thread
+                    THANDLE_MOVE(ASYNC_OP)(async_op_out, &async_op);
```

Comment:
Same comment like in ml, seems like the upper layer would not play nice if they do that.

Reply:
Added some notes in the MD and simplified the comment with best practice to maybe just return ASYNC_OP instead